### ISSUE TITLE
Research on pagination beyond last page

### ISSUE THREAD

**Issue by hcaofec** [fecgov/openFEC#4405]
User story:
From: Cody
My question: Are both of these key/value pairs necessary to get the next page? I had assumed that you needed both, but my script broke when one page of results included the last_index key/value pair but not the last_contribution_receipt_date key/value pair.
In other words, my script was doing fine with a last_indexes object that looks like this --
{'per_page': 100, 'last_indexes': {'last_index': '4051520201759281324', 'last_contribution_receipt_date': '2020-03-13T00:00:00'}, 'count': 706, 'pages': 8}
-- but choked when the last_indexes object looks like this --
{'pages': 8, 'count': 706, 'per_page': 100, 'last_indexes': {'last_index': '4032320051055495678', 'sort_null_only': True}}

--initial api call
https://api.open.fec.gov/v1/schedules/schedule_a/?api_key=DEMO_KEY&sort=contribution_receipt_date&sort_null_only=false&sort_hide_null=false&per_page=100&contributor_name=t+steyer&is_individual=true

pagination": {
    "count": 706,
    "per_page": 100,
    "last_indexes": {
      "last_index": "4081920091117540729",
      "last_contribution_receipt_date": "2009-03-31T00:00:00"
    },
    "pages": 8

query:
`select * from disclosure.fec_fitem_sched_a
where is_individual=true
and contributor_name_text @@ to_tsquery('T:* & STEYER:*')
;`

--Page 7--
https://api.open.fec.gov/v1/schedules/schedule_a/?api_key=DEMO_KEY&sort=contribution_receipt_date&sort_null_only=false&sort_hide_null=false&per_page=100&contributor_name=t+steyer&is_individual=true&last_index=4022620191643977930&last_contribution_receipt_date=2018-09-05T00:00:00

"pagination": {
    "per_page": 100,
    "last_indexes": {
      "last_index": "4051520201759281324",
      "last_contribution_receipt_date": "2020-03-13T00:00:00"
    },
    "count": 706,
    "pages": 8

query:
`select * from disclosure.fec_fitem_sched_a
where contb_receipt_dt IS NOT NULL
and is_individual=true
and contributor_name_text @@ to_tsquery('T:* & STEYER:*')
and (contb_receipt_dt, sub_id) > ('2020-03-13T00:00:00', 4022620191643977930)
;`

--Page 8 ( return last 6 rows, 3 rows with contribution_receipt_date = null )--
https://api.open.fec.gov/v1/schedules/schedule_a/?api_key=DEMO_KEY&sort=contribution_receipt_date&sort_null_only=false&sort_hide_null=false&per_page=100&contributor_name=t+steyer&is_individual=true&last_index=4051520201759281324&last_contribution_receipt_date=2020-03-13T00:00:00

pagination": {
    "pages": 8,
    "last_indexes": {
      "last_index": "4032320051055495678",
      "sort_null_only": true
    },
    "per_page": 100,
    "count": 706

--Page 9, return 502--
https://api.open.fec.gov/v1/schedules/schedule_a/?api_key=DEMO_KEY&sort=contribution_receipt_date&sort_null_only=true&sort_hide_null=false&contributor_name=t+steyer&is_individual=true&last_index=4032320051055495678

query:
`select * from disclosure.fec_fitem_sched_a
where contb_receipt_dt IS NULL
and is_individual=true
and contributor_name_text @@ to_tsquery('T:* & STEYER:*')
and sub_id > 4032320051055495678;`

--get value of last row--
`select max(sub_id), count(*) from disclosure.fec_fitem_sched_a
where is_individual=true
and contributor_name_text @@ to_tsquery('T:* & STEYER:*')
;`

**Comment by hcaofec** [comment]
Email 1:
Hello!
I am using the API to page through some schedule A donations and hit a snag. Per the API docs, I'm using the values in the last_indexes key to page through results:
To request the next page, you should append the values found in the last_indexes object from pagination to the URL of your last request as additional parameters.
In the example given, subsequent requests use the last_index and last_contribution_receipt_date values to get the next page of results.
My question: Are both of these key/value pairs necessary to get the next page? I had assumed that you needed both, but my script broke when one page of results included the last_index key/value pair but not the last_contribution_receipt_date key/value pair.
In other words, my script was doing fine with a last_indexes object that looks like this --
{'per_page': 100, 'last_indexes': {'last_index': '4051520201759281324', 'last_contribution_receipt_date': '2020-03-13T00:00:00'}, 'count': 706, 'pages': 8}
-- but choked when the last_indexes object looks like this --
{'pages': 8, 'count': 706, 'per_page': 100, 'last_indexes': {'last_index': '4032320051055495678', 'sort_null_only': True}}
-- because my script assumed the presence of both values. Would appreciate any help you can throw my way!
Thank you,
Cody Winchester

**Comment by hcaofec** [comment]
Email 2:
Hi Paul,
Thank you for this explanation. If I understand correctly, then, this request should return what I need? (I'm paging through schedule A donations from "T STEYER," at the moment.)
https://api.open.fec.gov/v1/schedules/schedule_a?
    api_key=<KEY>&
    contributor_name=t+steyer&
    per_page=100&
    sort=contribution_receipt_date&
    is_individual=true&
    last_index=4032320051055495678&
    sub_id=>4032320051055495678&
    sort_null_only=true
This just hangs for me, though -- no response for several minutes, so I'm assuming the request has failed. If I append last_contribution_receipt_data=null , I get a 422 code saying that null is not a valid date.
If helpful, this is the last_indexes object returned when I request the previous page of donation records: {'pages': 8, 'count': 706, 'per_page': 100, 'last_indexes': {'last_index': '4032320051055495678', 'sort_null_only': True}}
Thank you for any continued help you can throw my way! Happy to share my code (Python) too, if helpful.
Best,
Cody

**Comment by hcaofec** [comment]
Email 3: 
Hello!

Thank you for your response. Some quick back story and a code snippet:

I'm working with a journalist who needs to use the FEC API to help maintain a complete local database, updated once a month, of all federal donations by a group of newsworthy individuals (including Tom Steyer). So the goal is to run a script monthly to get a current, complete copy of all possible donations by these individuals, with on-the-ground reporting to verify that each donation was made by the individual in question and not by someone else with the same name.

Our strategy is:
Read in a CSV with her list of names
Search first initial + last name against the schedule_a endpoint, page through the donations returned while writing the data to file
Check/vet/dedupe/analyze/etc.
(We'd talked about using the bulk data files, but it looks like the API has more complete information, including <$200 donations and donor addresses, which she uses to help verify donor identity.)

Here's the bit of code in question. Still tied up on the logic of which params to include in subsequent calls based on the response. Thanks so much for taking a look!

Best,
Cody Winchester

**Comment by hcaofec** [comment]
User code:
https://gist.githubusercontent.com/cjwinchester/1e2423445edec3a84c1de36322c482c4/raw/09da65e75e63e3272d847a6b9355426101ec3e60/sked_a_donations.py

**Comment by fec-jli** [comment]
Add endpoint description about last index in /schedules/schedule_a/ endpoint. 
[https://github.com/fecgov/openFEC/pull/4426](https://github.com/fecgov/openFEC/pull/4426)
It is done, close this issue.
